src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/NuGetPowerShellUsageCollector.cs

```diff
@@ -0,0 +1,421 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Linq;
+using NuGet.Common;
+using NuGet.VisualStudio.Telemetry.Powershell;
+
+namespace NuGet.VisualStudio.Telemetry
+{
+    [Export(typeof(NuGetPowerShellUsageCollector))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public sealed class NuGetPowerShellUsageCollector : IDisposable
+    {
+        // PMC, PMUI powershell telemetry consts
+        public const string NuGetPMCExecuteCommandCount = nameof(NuGetPMCExecuteCommandCount);
+        public const string NuGetPMUIExecuteCommandCount = nameof(NuGetPMUIExecuteCommandCount);
+        public const string NuGetCommandUsed = nameof(NuGetCommandUsed);
+        public const string InitPs1LoadPMUI = nameof(InitPs1LoadPMUI);
+        public const string InitPs1LoadPMC = nameof(InitPs1LoadPMC);
+        public const string InitPs1LoadedFromPMCFirst = nameof(InitPs1LoadedFromPMCFirst);
+        public const string LoadedFromPMUI = nameof(LoadedFromPMUI);
+        public const string FirstTimeLoadedFromPMUI = nameof(FirstTimeLoadedFromPMUI);
+        public const string LoadedFromPMC = nameof(LoadedFromPMC);
+        public const string FirstTimeLoadedFromPMC = nameof(FirstTimeLoadedFromPMC);
+        public const string SolutionLoaded = nameof(SolutionLoaded);
+        public const string PowerShellExecuteCommand = nameof(PowerShellExecuteCommand);
+        public const string NuGetPowerShellLoaded = nameof(NuGetPowerShellLoaded);
+
+        // PMC UI Console Container telemetry consts
+        public const string PackageManagerConsoleWindowsLoad = nameof(PackageManagerConsoleWindowsLoad);
+        public const string NuGetPMCWindowLoadCount = nameof(NuGetPMCWindowLoadCount);
+        public const string ReOpenAtStart = nameof(ReOpenAtStart);
+
+        // Const name for emitting when VS solution close or VS instance close.
+        public const string Name = nameof(Name);
+        public const string NuGetPowershellPrefix = "NuGetPowershellPrefix."; // Using prefix prevent accidental same name property collission from different type telemetry.
+        public const string NuGetVSSolutionClose = nameof(NuGetVSSolutionClose);
+        public const string NuGetVSInstanceClose = nameof(NuGetVSInstanceClose);
```

Comment:
These two properties are being used as the telemetry event name. This class has a specific name, but the telemetry event name is very generic.  @dominoFire  do you have an opinion on what is better to query?  A single telemetry event with many, unrelated properties, or use the event name to partition different "families" of data, even when they are emit at the same time (solution close, VS close)?


We can refactor this class later to make it easier to re-use a generic event name later, but it will break all data queries if we change the event name, so that's more important to get right now.

Reply:
I consider it's better to have one event:

- Triggered when it's appropriate to make ordering assumptions, but, don't rely on timing data since each client/user/machine has their relative clocks
- Has all needed properties, to avoid further table joins